src/shared_key.rs

```diff
@@ -22,9 +25,217 @@ pub enum SharedKeyError {
     Generate,
 }
 
+const DITTO_IDENTITY_TAG: &str = "DITTO IDENTITY";
+
+pub const IN_BAND_CERTIFICATE_VERSION: u32 = 1;
+
+/// Information which is verified by the signature inside the `InBandCertificate`.
+#[derive(Debug, Serialize, Deserialize, Clone, PartialEq, Eq)]
+pub struct Payload {
+    /// Application for which this certificate is granted
+    #[serde(rename = "a")]
+    pub app_id: String,
+
+    /// Seconds since unix epoch that this certificate became valid, i.e. now() >= not_before
+    #[serde(rename = "n")]
+    pub not_before: i64,
+
+    /// Seconds since unix epoch until which this certificate remains valid, i.e. now() < expiry
+    #[serde(rename = "e")]
+    pub expiry: i64,
+
+    /// Hash of the serialised `IdentityData`, as it was issued for this peer upon authentication.
+    /// This is also a serialised Ditto TLV.
+    #[serde(rename = "h", with = "serde_bytes")]
+    pub identity_data_hash: Vec<u8>,
+}
+
+/// An assertion that a certain `IdentityData` is valid. The `IdentityData` contains a mapping from
+/// information about the peer to their Peer Key.
+#[derive(Debug, Serialize, Deserialize, Clone, PartialEq, Eq)]
+pub struct InBandCertificate {
+    /// Version of certificate format. See [IN_BAND_CERTIFICATE_VERSION].
+    #[serde(rename = "v")]
+    pub version: u32,
+
+    /// Signed content: a TLV-serialised form of [Payload].
+    ///
+    /// Stored as opaque binary to ensure the signature can be verified without decoding ambiguity.
+    #[serde(rename = "p", with = "serde_bytes")]
+    pub payload: Vec<u8>,
+
+    /// Public key of the signer. Should be 33 bytes.
+    #[serde(rename = "k", with = "serde_bytes")]
+    pub ecdsa_verifying_key: Vec<u8>,
+
+    /// ECDSA signature. Should be 64 bytes.
+    #[serde(rename = "s", with = "serde_bytes")]
+    pub ecdsa_signature: Vec<u8>,
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone, PartialEq, Eq)]
+pub struct InBandAuthV1 {
+    /// This peer's InBandCertificate
+    inband_certificate: InBandCertificate,
+    /// CA public keys to verify the in band certificate
+    inband_ca_pubkey_keys: Vec<Vec<u8>>,
+    // inband_metadata: Option<??>,
+    sub_authority_certificate: Option<Vec<u8>>,
```

Comment:
@thombles would this cover the sub authority cert & permissions info? Same question for the x509 below.

Reply:
Yep let's go with that for now. We can adjust it later if necessary.

The subauthority permissions are in the identity data so that's included automatically. We just have a different cert when acting as an intermediate authority vs as a end-user peer.